ngraph/core/reference/include/ngraph/runtime/reference/convolution_backprop_data.hpp

```diff
@@ -23,217 +19,224 @@ namespace ngraph
+                                                }
+                                            }
+                                        }
+
+                                        if (i_y < input_3d[1] - 1)
+                                        {
+                                            for (int y_dim = 0;
+                                                 y_dim < new_input_3d[2] * (strides_3d[1] - 1);
+                                                 y_dim++)
+                                            {
+                                                input_zeros.push_back(0);
+                                            }
+                                        }
+                                    }
+
+                                    if (i_z < input_3d[0] - 1)
+                                    {
+                                        for (int y_dim = 0;
+                                             y_dim < new_input_3d[1] * new_input_3d[2] *
+                                                         (strides_3d[0] - 1);
+                                             y_dim++)
+                                        {
+                                            input_zeros.push_back(0);
+                                        }
+                                    }
+                                }
+                            }
+                        }
                     }
+                }
+            } // namespace
 
-                    CoordinateTransform filter_transform(
-                        filter_shape, filter_transform_start, filter_transform_end);
+            template <typename T>
+            void convolution_backprop_impl(const T* in,
+                                           const T* f,
+                                           T* out,
+                                           const Shape& in_shape,
+                                           const Shape& f_shape,
+                                           const Shape& out_shape,
+                                           const Strides& strides,
+                                           const Strides& dilation,
+                                           const CoordinateDiff& pads_begin,
+                                           const CoordinateDiff& pads_end,
+                                           const CoordinateDiff& output_padding)
 
-                    // As we go, we sum up:
-                    //
-                    //   out[O] += in[I] * filter[F].
+            {
+                // this implementation supports 1D, 2D and 3D convolutions
+                NGRAPH_CHECK(in_shape.size() >= 3 && in_shape.size() <= 5,
+                             "Unsupported input rank: ",
+                             in_shape);
+
+                NGRAPH_CHECK(f_shape.size() >= 3 && f_shape.size() <= 5,
+                             "Unsupported kernel rank: ",
+                             f_shape);
```

Comment:
When using dynamic shapes, checks in `validate_and_infer_types` for correctness of shapes are by-passed. So that then at runtime we may have incorrect/incompatible shapes. We should add checks here before we jump to calculations. Same as in convolution reference implementation: [link](https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/convolution.hpp#L182)